resource/schema/infra_data.cpp

```diff
@@ -56,62 +56,94 @@ pool_infra_t::pool_infra_t ()
         if (!p)
             continue;
-        base_time = planner_multi_base_time (p);
-        duration = planner_multi_duration (p);
-        size_t len = planner_multi_resources_len (p);
-        subplans[kv.first] = planner_multi_new (base_time, duration,
-                                 planner_multi_resource_totals (p),
-                                 planner_multi_resource_types (p), len);
+        subplans[kv.first] = planner_multi_copy (p);
     }
     if (o.x_checker) {
-        base_time = planner_base_time (o.x_checker);
-        duration = planner_duration (o.x_checker);
-        x_checker = planner_new (base_time, duration,
-                                 planner_resource_total (o.x_checker),
-                                 planner_resource_type (o.x_checker));
+        if (!x_checker) {
+            x_checker = planner_copy (o.x_checker);
+        } else {
+            *x_checker = *(o.x_checker);
+        }
     }
 }
 
 pool_infra_t &pool_infra_t::operator= (const pool_infra_t &o)
 {
-    int64_t base_time = 0;
-    uint64_t duration = 0;
+    scrub ();
+    subplans.clear ();
 
-    // don't copy the content of infrastructure tables and subtree
-    // planner objects.
     infra_base_t::operator= (o);
+    ephemeral = o.ephemeral;
     colors = o.colors;
+    tags = o.tags;
+    x_spans = o.x_spans;
+    job2span = o.job2span;
+
     for (auto &kv : o.subplans) {
         planner_multi_t *p = kv.second;
         if (!p)
             continue;
-        base_time = planner_multi_base_time (p);
-        duration = planner_multi_duration (p);
-        size_t len = planner_multi_resources_len (p);
-        subplans[kv.first] = planner_multi_new (base_time, duration,
-                                 planner_multi_resource_totals (p),
-                                 planner_multi_resource_types (p), len);
+        subplans[kv.first] = planner_multi_copy (p);
     }
     if (o.x_checker) {
-        base_time = planner_base_time (o.x_checker);
-        duration = planner_duration (o.x_checker);
-        x_checker = planner_new (base_time, duration,
-                                 planner_resource_total (o.x_checker),
-                                 planner_resource_type (o.x_checker));
+        if (!x_checker) {
+            x_checker = planner_copy (o.x_checker);
+        } else {
+            *x_checker = *(o.x_checker);
+        }
     }
     return *this;
 }
 
+bool pool_infra_t::operator== (const pool_infra_t &o) const
+{
+    if (tags != o.tags)
+        return false;
+    if (x_spans != o.x_spans)
+        return false;
+    if (job2span != o.job2span)
+        return false;
+    for (auto const &this_it : colors) {
+        auto const other = o.colors.find (this_it.first);
+        if (other == o.colors.end ())
+            return false;
+        if (this_it.first != other->first)
+            return false;
+        if (this_it.second != other->second)
+            return false;
+    }
```

Comment:
I'm relatively sure this is equivalent, if so it's a bit better self-documenting and somewhat faster (avoiding the O(logN) search on each element):

```suggestion
    if(colors.size() != o.colors.size()
       || !std::equal(this.colors.begin(), this.colors.end(), o.colors.begin()))
       return false;
```

Makes me want ranges, very badly, and autogenerated `<==>`, _sigh_.  Either way, same below.

Reply:
Actually, why is this one being compared differently than tags? The same `==` should work here as well for complete equality right?

Reply:
Actually, yes, `if (colors != o.colors)` works, too.

Regarding comparing `subplans`, this strategy doesn't work, because it's a map of `planner_multi_t` which is a wrapper struct around the `planner_multi` class. Adding an `==` operator to `planner_multi_t` doesn't work with `std::equal`, unfortunately.